lib/iris/fileformats/cf.py

```diff
@@ -1295,7 +1295,10 @@ def _reset(self):
 
     def __del__(self):
         # Explicitly close dataset to prevent file remaining open.
-        self._dataset.close()
+        try:
+            self._dataset.close()
+        except AttributeError:
```

Comment:
TBH I think you should ignore virtually **any** error here.
I can't think of anything it would be sensible to raise here, since there is no code context for what happened (just the garbage collector).

It's also fair to say that this belt+braces is a possibly bit odd, and it might be more logical to trust the dataset (or underlying file) to tidy up when _it_ is deferenced or garbage-collected.
Maybe there was at some time a problem with the netCDF4.Dataset that necessitated this, but I bet there isn't any more ?

Reply:
@pp-mo  That was initially my thought as well. See https://github.com/SciTools/iris/issues/3312#issuecomment-500452299 . See bjlittle's response below my comment, which is an explanation of why `__del__` is there.

Reply:
Ok, but in that case we maybe need to be clear why we are swallowing some errors but not others.
It seems to be simple a bug in the CFReader -- it doesn't always have a `_dataset`.

So I think it would make far more sensible + robust to init `self._dataset = None` + test before closing.
[Like this]() ??

Reply:
That's what I had in [this commit](https://github.com/SciTools/iris/pull/4646/commits/0bf13f4825cbc69339c0c6c3ddc1fd15a3366288) and @evertrol had in their original PR (#3336) - I just thought it more pythonic to ask forgiveness rather than permission. I can put it back

Reply:
Then we're back at https://github.com/SciTools/iris/pull/3336 (which I closed for a different reason).

I'm still in favour of removing `__del__`, and let the garbage collector handling the destruction and closing of the dataset. If `netCDF4.Dataset` doesn't close its file pointer upon garbage collection/destruction, then that is the thing that should be fixed, imo.
Then again, I see no `__del__` method in `netCDF4.Dataset`, while `close()` does a lot of things there (there are also `__enter__` and `__exit__` methods, but these are not applicable here). But this related [netCDF4-python issue](https://github.com/Unidata/netcdf4-python/issues/218) does mention `__dealloc__` takes of the file closing. 
So once there are no references to the dataset, the underlying file get closed. Which suggests that letting Python garbage collect the CFReader instance, and thereby also its `_dataset` attribute, will close the file.

But someone else should decide; in particular as I haven't used Scitools for well over a year.